event/event.go

```diff
@@ -204,6 +204,8 @@ type SyncEventMetadata struct {
 	// policy changes, and "other" (meaning things we didn't commit
 	// ourselves)
 	Includes map[string]bool `json:"includes,omitempty"`
+	// Per-resource errors
+	Errors map[string]string `json:"errors,omitempty"`
```

Comment:
Can we give this a better type, perhaps? `map[T]error` where `T` is some meaningful struct type with a good `String()` implementation would help make this self-documenting.

I think the `ResourceID` code is a good example of making something that _is_ stringy not be so stringy.

Reply:
Yes, probably just 

```
type ErrorSource struct {
    ResourceID string
    Source string
}
```

(those are the two bits of information that get encoded in here). I don't think we much care about unpacking it at the other side.

Reply:
Or,

```
type ResourceError struct {
  ID string
  Path string
  Error string
}

type SyncEventMetadata struct {
...
  Errors []ResourceError
}
```

Reply:
The intended result here is that consumers of this struct (ui, notifications) don't need to search through the code to find out what kind of string we build and put in here.

The most-general, least-informative types we can JSONify are `map[string]string` and `map[string]interface{}`, both of which are basically `Object`. In this case the `V` in our `map[K]V` can't really be narrowed down to more than just arbitrary error text, but the `K` certainly can be.

If there's a way we can't build a key from `resource.Resource` that'd probably be perfect :)

Reply:
I'd rather have a slice of structs than have to pack things into a string so it can be a JSON-able map.

Reply:
Indeed if we don't actually need a map, let's scrap the map! Now let's figure out what a `struct { string, string, string }` needs... 😄 

(`ResourceError` up there doesn't have the same issue because it has named fields, but maybe we can give `ID` similar treatment to the map keys. `resource.ID`?)

Reply:
> maybe we can give ID similar treatment to the map keys

At some point we have to bottom out to JSON types. This is that point.

Reply:
Technically we could probably use a `ResourceID` there, since it'll get serialised to a string anyway.